src/libraries/System.Private.CoreLib/src/System/String.Comparison.cs

```diff
@@ -834,43 +836,108 @@ internal unsafe int GetNonRandomizedHashCode()
-                // be ok because we expect this to be very rare in practice.
+            // We "normalize to lowercase" every char by ORing with 0x0020. This casts
+            // a very wide net because it will change, e.g., '^' to '~'. But that should
+            // be ok because we expect this to be very rare in practice.
 
-                const uint NormalizeToLowercase = 0x0020_0020u; // valid both for big-endian and for little-endian
+            const uint NormalizeToLowercase = 0x0020_0020u; // valid both for big-endian and for little-endian
 
-                while (length > 2)
+            nint i = 0;
+            int count = length;
+
+            if (normalizeNonAscii)
+            {
+                while (count > 2)
                 {
-                    length -= 4;
-                    // Where length is 4n-1 (e.g. 3,7,11,15,19) this additionally consumes the null terminator
-                    hash1 = (BitOperations.RotateLeft(hash1, 5) + hash1) ^ (ptr[0] | NormalizeToLowercase);
-                    hash2 = (BitOperations.RotateLeft(hash2, 5) + hash2) ^ (ptr[1] | NormalizeToLowercase);
-                    ptr += 2;
+                    ref byte firstByte = ref Unsafe.As<char, byte>(ref firstChar);
+                    uint p0 = Unsafe.ReadUnaligned<uint>(ref Unsafe.Add(ref firstByte, i));
+                    uint p1 = Unsafe.ReadUnaligned<uint>(ref Unsafe.Add(ref firstByte, i + 4));
+                    if (!Utf16Utility.AllCharsInUInt32AreAscii(p0 | p1))
+                    {
+                        goto NotAscii;
+                    }
+
+                    // Where count is 4n-1 (e.g. 3,7,11,15,19) this additionally consumes the null terminator
+                    hash1 = (BitOperations.RotateLeft(hash1, 5) + hash1) ^ (p0 | NormalizeToLowercase);
+                    hash2 = (BitOperations.RotateLeft(hash2, 5) + hash2) ^ (p1 | NormalizeToLowercase);
+
+                    count -= 4;
+                    i += 8;
                 }
 
-                if (length > 0)
+                if (count > 0)
                 {
-                    // Where length is 4n-3 (e.g. 1,5,9,13,17) this additionally consumes the null terminator
-                    hash2 = (BitOperations.RotateLeft(hash2, 5) + hash2) ^ (ptr[0] | NormalizeToLowercase);
+                    uint p0 = Unsafe.ReadUnaligned<uint>(ref Unsafe.Add(ref Unsafe.As<char, byte>(ref firstChar), i));
+                    if (!Utf16Utility.AllCharsInUInt32AreAscii(p0))
+                    {
+                        goto NotAscii;
+                    }
+
+                    // Where count is 4n-3 (e.g. 1,5,9,13,17) this additionally consumes the null terminator
+                    hash2 = (BitOperations.RotateLeft(hash2, 5) + hash2) ^ (p0 | NormalizeToLowercase);
+                }
+            }
+            else
+            {
+                // The duplication can be removed once JIT gets "Loop Unswitching" optimization
```

Comment:
Until that happens, this code can be in the slow path to make the whole thing a bit more efficient.